Sources/ServiceLifecycle/GracefulShutdown.swift

```diff
@@ -0,0 +1,154 @@
+/// Execute an operation with a graceful shutdown handler that’s immediately invoked if the current task is shutting down gracefully.
+///
+/// This doesn’t check for graceful shutdown, and always executes the passed operation.
+/// The operation executes on the calling execution context and does not suspend by itself, unless the code contained within the closure does.
+/// If graceful shutdown occurs while the operation is running, the graceful shutdown handler will execute concurrently with the operation.
+///
+/// When `withShutdownGracefulHandler` is used in a Task that has already been gracefully shutdown, the `onGracefulShutdown` handler
+/// will be executed immediately before operation gets to execute. This allows the `onGracefulShutdown` handler to set some external “shutdown” flag
+/// that the operation may be atomically checking for in order to avoid performing any actual work once the operation gets to run.
+///
+/// A common use-case is to listen to graceful shutdown and use the `ServerQuiescingHelper` from `swift-nio-extras` to
+/// trigger the quiescing sequence. Furthermore, graceful shutdown will propagate to any child task that is currently executing
+///
+/// - Parameters:
+///   - operation: The actual operation.
+///   - handler: The handler which is invoked once graceful shutdown has been triggered.
+@_unsafeInheritExecutor
+public func withShutdownGracefulHandler<T>(
+    operation: () async throws -> T,
+    onGracefulShutdown handler: @Sendable @escaping () -> Void
```

Comment:
I think you should keep the state machine. It is in general just good to have that to reduce state handling error.

W.r.t. to async or not for the closure. I followed the pattern that onCancel established but we could also make it async. Open for input here.

I understand that this would result in another unstructured Task to communicate to your actor otherwise.
As a general note, the onGracefulShutdown is not running on your actor executor but on the one calling graceful shutdown. So you still need to hop on your actor before doing state modifications

Reply:
Having thought about this again. I think `onGracefulShutdown` should be non `async`. The reason for this is that we should follow the pattern of how cancellation works. This is due to the fact on how actually triggering graceful shutdown also works. If we call `GracefulShutdownManager.shutdownGracefully()` we can't be sure that the tasks actually shutdown after that call. We can only be sure that it finished once the child tasks return.

Probably the thing we want to do here is use and unstructured task in the `onGracefulShutdown` closure and call back to the actor server which will then do the right thing. In the future, we might be able to avoid that unstructured task by having `send` semantics for actor method invocations cc @ktoso

Reply:
So are you saying we shouldn't make `onGracefulShutdown` async because it might cause confusion and imply that once it returns it has completed shutdown?

It seems odd though that everything else about this module points the user to use structured concurrency, but as soon as we get to shutdown the recommendation is use an unstructured task.

Reply:
Fully agree with you; however, this is just a missing thing in the language IMO. We already have this problem with `onCancel` and `onGracefulShutdown` is just going to add to that list. Actors should provide a way to send them a message that is ordered correctly and doesn't await a response.

I wouldn't want to make the handler `async` to just support that case.